src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -27,11 +32,78 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepComp
             // name binding is done
             // allow type queries now
             paramsSymbolContext.Unlock();
+            // lazy load single use diagnostic set
+            this.AllDiagnostics = new Lazy<ImmutableArray<IDiagnostic>>(() => GetDiagnostics());
         }
 
-        public IEnumerable<IDiagnostic> GetDiagnostics()
-            => BicepParamFile.ProgramSyntax.GetParseDiagnostics();
+        /// <summary>
+        /// Gets all the params semantic diagnostics unsorted. Does not include params parser and params lexer diagnostics.
+        /// </summary>
+        /// <returns></returns>
+        private IEnumerable<IDiagnostic> GetAdditionalSemanticDiagnostics()
+        {
+            if (this.BicepCompilation is null)
+            {
+                return Enumerable.Empty<IDiagnostic>();
+            }
+
+            var diagnosticWriter = ToListDiagnosticWriter.Create();
+
+            var parameters = this.BicepCompilation.GetEntrypointSemanticModel().Parameters;
+            var parameterAssignments = BicepParamFile.ProgramSyntax.Children.OfType<ParameterAssignmentSyntax>().Where(x => this.ParamBinder.GetSymbolInfo(x) is not null);
+
+            // get diagnostics relating to missing parameter assignments or declarations
+            GetParameterMismatchDiagnostics(diagnosticWriter, parameters, parameterAssignments);
+            // get diagnostics relating to type mismatch of params between Bicep and params files
+            GetTypeMismatchDiagnostics(diagnosticWriter, parameterAssignments);
+
+            return diagnosticWriter.GetDiagnostics();
+        }
+        
+        private void GetParameterMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, ImmutableArray<Metadata.ParameterMetadata> parameters, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
+        {
+            var parametersDict = parameters.ToDictionary(x => x.Name, LanguageConstants.IdentifierComparer);
+            var parametersAssignmentsDict = parameterAssignments.ToDictionary(x => x.Name.IdentifierName, LanguageConstants.IdentifierComparer);
+            
+            // parameters that are declared but not assigned
+            var missingRequiredParams = parametersDict.Values
+                .Where(x => x.IsRequired)
+                .Where(x => !parametersAssignmentsDict.ContainsKey(x.Name))
+                .OrderBy(x => x.Name);
+            // parameters that are assigned but not declared
+            var missingAssignedParams = parametersAssignmentsDict
+                .Where(x => !parametersDict.ContainsKey(x.Key))
+                .Select(x => x.Value)
+                .OrderBy(x => x.Name.IdentifierName);
+            
+            foreach (var requiredParam in missingRequiredParams)
+            {
+                diagnosticWriter.Write(new TextSpan(0, 0), x => x.MissingParameterAssignment(requiredParam.Name));
```

Comment:
[nit] Have you discussed the pros/cons of showing a diagnostic per missing required parameter vs a single diagnostic containing all the missing required parameters? Personally, I'm not sure what the best option is, but worth thinking about as I expect it'll be the first thing that many users will see after they fill out their `using` statement.

Here's a somewhat similar diagnostic we have for missing required params in a `module` declaration, just to give an idea of what a combined diagnostic would look like:
<img width="612" alt="image" src="https://user-images.githubusercontent.com/38542602/180945945-143ca762-faa6-40f8-8b8d-8304ea7e4d52.png">

Another thing perhaps worth discussing w.r.t this entry-point user experience - pros/cons of putting missing params diagnostics on the `using` statement instead of at position 0,0.

Finally (and totally out of scope for this PR), it would be **very cool** to have a code fix to insert missing param declarations, with tabstops on them - similar to this experience:

https://user-images.githubusercontent.com/38542602/180947746-318031e0-455d-43a5-9d73-4bf17e53782c.mov

Reply:
I like both ideas!

Reply:
re: per param vs. single... the more actionable the error message is the better.  if the single contains all that are missing/invalid that works but is probably unwieldy until the "quick fix" is implemented.

Reply:
Personally I'd put them into a single error since: a) the location will be the exact same place, b) it's what Bicep does:
<img width="951" alt="image" src="https://user-images.githubusercontent.com/6913354/181101790-389ded2e-9dfa-4be6-b74b-41abb89f7b9f.png">

But I don't think it's a big deal, and it can always be changed afterwards according to feedback.

Reply:
I'm going to go ahead and merge this PR (once it passes all the tests 😭), but we should definitely discuss this next week!